Add optimistic locking support to sqlite backend

rosstimothy opened the pull request:
A new revision column is added to the `kv` table to accomodate the storing the resource revision. A schema migration for the table was added to ensure any existing databases have the table added. Additionaly the `meta` table and its associated import functionality was removed since it was unused. Various other small clean up operations were also done to eliminate warnings and remove unused functionality.

Contributes to #30416.

espadolini commented on lib/backend/lite/lite.go:
```diff
@@ -476,12 +491,11 @@ func (l *Backend) Put(ctx context.Context, i backend.Item) (*backend.Lease, erro
-		}
+		recordID := id(created)
 		if !l.EventsOff {
 			stmt, err := tx.PrepareContext(ctx, "INSERT INTO events(type, created, kv_key, kv_modified, kv_expires, kv_value) values(?, ?, ?, ?, ?, ?)")
```
Don't we need the revision in the events, too?

espadolini commented on lib/backend/lite/lite.go:
```diff
@@ -337,16 +330,30 @@ func (l *Backend) createSchema() error {
+	}
 
-func (l *Backend) newLease(item backend.Item) *backend.Lease {
-	var lease backend.Lease
-	if item.Expires.IsZero() {
-		return &lease
+	columns := make(map[string]struct{})
+	for rows.Next() {
+		var tmp, name, null sql.NullString
+		err := rows.Scan(&tmp, &name, &tmp, &null, &tmp, &tmp)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		columns[name.String] = struct{}{}
+	}
+
+	if _, ok := columns["revision"]; !ok {
+		const addRevision = `ALTER TABLE kv ADD COLUMN revision TEXT NOT NULL DEFAULT "";`
+		if _, err := l.db.ExecContext(l.ctx, addRevision); err != nil {
+			l.Errorf("Failing schema step: %v, %v.", addRevision, err)
+			return trace.Wrap(err)
+		}
 	}
-	lease.Key = item.Key
-	return &lease
+
+	return nil
```
If we're not ok with just attempting to add the column and ignoring the error if the column exists, why not do `select exists(select 1 from pragma_table_info('kv') where name = 'revision');` rather than parsing the entire table info?

espadolini commented on lib/backend/lite/lite.go:
```diff
@@ -887,6 +785,92 @@ func (l *Backend) DeleteRange(ctx context.Context, startKey, endKey []byte) erro
+	err := l.inTransaction(ctx, func(tx *sql.Tx) error {
+		now := l.clock.Now().UTC()
+
+		stmt, err := tx.PrepareContext(ctx, "UPDATE kv SET value = ?, expires = ?, modified = ?, revision = ? WHERE key = ? AND revision = ?")
```
We should check that the existing value isn't expired here; the current write operations (other than `Update`) do the same.

espadolini commented on lib/backend/lite/lite.go:
```diff
@@ -887,6 +785,92 @@ func (l *Backend) DeleteRange(ctx context.Context, startKey, endKey []byte) erro
+	}
+
+	return l.inTransaction(ctx, func(tx *sql.Tx) error {
+		stmt, err := tx.PrepareContext(ctx, "DELETE FROM kv WHERE key = ? AND revision = ?")
```
I don't know how to feel about returning no error when deleting an expired item, especially with a conditional operation. The memory backend that was just merged will error out in that case, however.

espadolini commented on lib/backend/lite/lite.go:
```diff
@@ -1000,55 +984,33 @@ func isClosedError(err error) bool {
-	e, ok := trace.Unwrap(err).(sqlite3.Error)
-	if !ok {
+	var e sqlite3.Error
+	if ok := errors.As(trace.Unwrap(err), &e); !ok {
```
Won't `errors.As` call Unwrap on the source error?

espadolini commented on lib/backend/lite/lite.go:
```diff
@@ -337,16 +332,43 @@ func (l *Backend) createSchema() error {
+		}
+
+		if migrationRequired {
+			query := fmt.Sprintf(`ALTER TABLE %s ADD COLUMN %s TEXT NOT NULL DEFAULT "";`, table, column)
```
Do we have a general idea of how long this will take?

I also just realized that we should do this in a read-write transaction so other instances of the backend (if any) are not going to get confused as they first check that the column doesn't exist and then fail to add a column that was added by someone else.

espadolini commented on lib/backend/lite/lite.go:
```diff
@@ -320,14 +317,12 @@ func (l *Backend) createSchema() error {
            kv_expires DATETIME,
-           kv_value BLOB
+           kv_value BLOB,
+           kv_revision TEXT
```
```suggestion
           kv_revision TEXT NOT NULL
```

espadolini replied:
I wasn't thinking about that, fair point.

(we could have the revision of the deleted item, it's just not useful)

espadolini replied:
I don't really like the inconsistency between the freshly created table and the table with the added columns tho, can we change both this and `revision` in `kv` to `TEXT NOT NULL DEFAULT ""` to match the column after `migrateRevision`, or have an extra parameter for the column constraint in `migrateRevision`?

espadolini commented on lib/backend/lite/lite.go:
```diff
@@ -771,23 +684,23 @@ func (l *Backend) KeepAlive(ctx context.Context, lease backend.Lease, expires ti
 			defer stmt.Close()
 
-			if _, err := stmt.ExecContext(ctx, types.OpPut, created, string(item.Key), id(created), expires.UTC(), item.Value); err != nil {
+			if _, err := stmt.ExecContext(ctx, types.OpPut, created, string(item.Key), id(created), expires.UTC(), item.Value, item.Revision); err != nil {
```
`item.Revision` should be updated with a new value before this, and the same one should be used for the events and the kv table.

espadolini commented on lib/backend/lite/lite.go:
```diff
@@ -887,6 +800,92 @@ func (l *Backend) DeleteRange(ctx context.Context, startKey, endKey []byte) erro
+	rev := backend.CreateRevision()
+	err := l.inTransaction(ctx, func(tx *sql.Tx) error {
+		now := l.clock.Now().UTC()
+		stmt, err := tx.PrepareContext(ctx, "UPDATE kv SET value = ?, expires = ?, modified = ?, revision = ? WHERE key = ? AND revision = ? AND (expires IS NULL OR expires > ?)")
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		defer stmt.Close()
+
+		result, err := stmt.ExecContext(ctx, i.Value, expires(i.Expires), id(now), rev, string(i.Key), i.Revision, now)
```
I don't think we get anything out of a separate prepare and exec step, if done like this, so we can just `tx.ExecContext()` instead.

rosstimothy committed:
Add optimistic locking support to sqlite backend

A new revision column is added to the kv table to accomodate
storing the resource revision. A schema migration for the table
was added to ensure any existing databases have the table added.
Additionaly the meta table and its associated import functionality
was removed since it was unused. Various other small clean up
operations were also done to eliminate warnings and remove unused
functionality.

Contributes to #30416.
